File: Packages/TypeEngine/Sources/TypeEngine/BeamDecoder.swift
import Foundation
import Lexicon

// MARK: - Per-position cost seam

/// Per-position character-level likelihood seam for the beam decoder.
///
/// The decoder prices every hypothesized "the user tapped `typed` but meant
/// `intended`" substitution through this protocol instead of calling the
/// static `SpatialModel` directly, because coordinate evidence will later
/// flow in PER TAP (PLAN.md "Touch decoding", stage 1) — and it works both
/// directions:
///
///  * a tap that landed near a key BOUNDARY makes the neighbor substitution
///    cheap (enables corrections the static model prices at ~1 nat anyway),
///  * a tap that landed DEAD CENTER makes every substitution expensive — a
///    correction veto: believe the user, they hit exactly what they meant.
///
/// `confidence(position:)` exists for the same wave: the corrector will
/// aggregate per-word tap confidence to modulate the autocorrect margin
/// (a word typed with all-center taps should almost never auto-replace).
/// Today the only concrete provider is the static keyboard-geometry one
/// below, whose confidence is a constant — the seam exists so the
/// coordinate-plumbing wave swaps providers without touching the beam.
public protocol PositionCostProvider {
    /// -log P(typed char | intended char) at this typed-character position,
    /// in nats. 0 when equal.
    func substitutionCost(position: Int, typed: Character, intended: Character) -> Double
    /// How confidently the tap at `position` selects its resolved character,
    /// in [0, 1]. Constant 1 until per-tap coordinates flow.
    func confidence(position: Int) -> Double
}

/// The static provider: key-center Gaussian distances from `SpatialModel`,
/// identical at every position; no per-tap evidence, so confidence is 1.
struct StaticSpatialCostProvider: PositionCostProvider {
    let spatial: SpatialModel

    func substitutionCost(position: Int, typed: Character, intended: Character) -> Double {
        spatial.substitutionCost(typed: typed, intended: intended)
    }

    func confidence(position: Int) -> Double { 1.0 }
}

// MARK: - Beam decoder

/// Beam-search spatial decoder: finds lexicon words reachable from the typed
/// token by spatially-plausible edits, by walking typed characters and a
/// lexicon prefix cursor in lockstep (uniform-cost search over
/// (prefix range, chars consumed) states).
///
/// This replaces the edits1/edits2 generate-and-test passes as the PRIMARY
/// candidate source: instead of materializing ~500–150k candidate strings
/// and probing each against the lexicon (~30µs a probe — the koetip 31ms
/// landmine), the search only ever visits prefixes the lexicon actually
/// contains, so multi-position adjacent-key noise ("koetip" → "kortið",
/// two 1-nat substitutions) is found in well under a millisecond.
///
/// Walk representation: per typed CHARACTER (extended grapheme cluster; for
/// this alphabet always a single scalar, 1–2 UTF-8 bytes). Every descend
/// appends one whole character's UTF-8 to the byte-sorted pool cursor, so
/// cursors stay scalar-aligned by construction — the byte-sortedness of the
/// pool is an implementation detail the decoder never observes partially.
///
/// Ops per state (all costs in nats, same currency as `SpatialModel`):
///  * match          — consume typed char, descend by it, cost 0
///  * substitution   — consume typed char, descend by a NEIGHBOR character,
///                     priced per position by the `PositionCostProvider`;
///                     the first edit may substitute ANY alphabet character
///                     (parity with the old edits1), later edits only the
///                     precomputed cheap-neighbor set (≤ beamNeighborMaxCost
///                     from static key geometry — a slightly generous
///                     superset that stays valid when coordinate evidence
///                     shifts costs around a tap point, since taps are
///                     spatially local)
///  * transposition  — consume two typed chars, descend in swapped order
///  * deletion       — consume typed char without descending (user typed an
///                     unintended extra character; `Costs.insertion`)
///  * insertion      — descend without consuming (user omitted an intended
///                     character; `Costs.deletion`), including after the
///                     last typed char
///
/// States dedup on (range lower bound, byte depth, chars consumed) — that
/// triple identifies the prefix — keeping the min accumulated cost, so the
/// emitted cost is the minimum over alignments within the edit/cost bounds.
/// Uniform-cost order means candidates emit cheapest-first: budget or cap
/// exhaustion sheds only the least plausible tail.
final class BeamDecoder {
    let config: EngineConfig
    /// Full candidate alphabet with precomputed UTF-8 (insertions + the
    /// first edit's substitutions).
    private let alphabet: [(char: Character, bytes: [UInt8])]
    /// Cheap spatial neighbors per alphabet character (static geometry:
    /// keys within `beamNeighborMaxCost` nats ≈ 1.5 key widths, accent
    /// twins, orthographic-confusion pairs), for second/third edits.
    private let cheapNeighbors: [Character: [(char: Character, bytes: [UInt8])]]

    /// Memo for SHALLOW descends (byte depth ≤ 2 at the parent). The first
    /// two levels of the walk sit on huge ranges (the whole table, then a
    /// first-letter bucket), so their binary searches are the widest — and
    /// they repeat identically across states, keystrokes and words. The
    /// memo is bounded by construction (≤ alphabet² entries per level per
    /// lexicon) and lives for the corrector's lifetime; correctness is
    /// unaffected because descend is a pure function of (lexicon, cursor,
    /// bytes). Keyed per lexicon identity index (0 = IS, 1 = EN).
    private struct ShallowKey: Hashable {
        let lexicon: Int
        let lowerBound: Int
        let byteDepth: Int
        let char: Character
    }
    private var shallowCache: [ShallowKey: LexiconPrefixCursor] = [:]

    /// Ranges at or below this size are expanded by scanning their actual
    /// children (`childCursors`) instead of probing per character.
    static let childScanLimit = 48

    init(config: EngineConfig, spatial: SpatialModel) {
        self.config = config
        var alphabet: [(char: Character, bytes: [UInt8])] = []
        for ch in Corrector.alphabet {
            alphabet.append((ch, Array(String(ch).utf8)))
        }
        self.alphabet = alphabet
        var neighbors: [Character: [(char: Character, bytes: [UInt8])]] = [:]
        for typed in Corrector.alphabet {
            neighbors[typed] = alphabet.filter { entry in
                entry.char != typed
                    && spatial.substitutionCost(typed: typed, intended: entry.char)
                        <= config.beamNeighborMaxCost
            }
        }
        self.cheapNeighbors = neighbors
    }

    private struct State {
        var cursor: LexiconPrefixCursor
        var consumed: Int
        var cost: Double
        var edits: Int
    }

    private struct VisitKey: Hashable {
        let lowerBound: Int
        let byteDepth: Int
        let consumed: Int
    }

    /// Decode `typed` against one lexicon. Returns up to
    /// `beamMaxCandidates` (word, accumulated channel cost) pairs, cheapest
    /// first. The caller re-scores with the exact DP (`Corrector
    /// .spatialCost`) — the beam cost is a search/pruning currency only.
    /// `pricing` is the lane-relaxation layer (PLAN.md "Lane relaxation
    /// profiles"): fold-pair substitutions and apostrophe insertions are
    /// priced min(provider, lane fold cost), composing ON TOP of the
    /// per-position provider seam — the future per-tap provider swap and
    /// the lane layer stay orthogonal (per-tap confidence will multiply
    /// into fold pricing, not replace it; see `FoldPricing`).
    func decode(
        typed: [Character],
        lexicon: PrefixSearchableLexicon,
        lexiconIndex: Int,
        costs positionCosts: PositionCostProvider,
        pricing: FoldPricing,
        maxEdits: Int
    ) -> [(word: String, cost: Double)] {
        let n = typed.count
        guard n > 0, maxEdits >= 0 else { return [] }
        let start = ContinuousClock.now
        let costCap = config.beamCostCap
        let insertionCost = config.spatialCosts.deletion  // omitted intended char
        let deletionCost = config.spatialCosts.insertion  // extra typed char
        let transpositionCost = config.spatialCosts.transposition
        let typedBytes: [[UInt8]] = typed.map { Array(String($0).utf8) }

        /// Provider cost with the lane-relaxation layer on top (see the
        /// `pricing` parameter doc).
        func substitutionPrice(position: Int, typedChar: Character, intended: Character) -> Double {
            let base = positionCosts.substitutionCost(
                position: position, typed: typedChar, intended: intended)
            guard
                let lane = pricing.substitutionPrice(
                    typed: typedChar, intended: intended, confusionBase: base)
            else { return base }
            return min(lane, base)
        }

        /// Descend with the shallow-level memo (see `shallowCache`).
        func descend(
            _ cursor: LexiconPrefixCursor, char: Character, bytes: [UInt8]
        ) -> LexiconPrefixCursor {
            // Memo key soundness: distinct NON-empty prefix cursors at the
            // same depth occupy disjoint ranges, so (depth, lowerBound)
            // identifies the prefix; empty cursors (which could alias a
            // neighbor's lowerBound) are never memoized — expansion only
            // descends from non-empty cursors anyway.
            guard cursor.byteDepth <= 2, !cursor.isEmpty else {
                return lexicon.descend(cursor, appendingUTF8: bytes)
            }
            let key = ShallowKey(
                lexicon: lexiconIndex,
                lowerBound: cursor.lowerBound,
                byteDepth: cursor.byteDepth,
                char: char
            )
            if let cached = shallowCache[key] { return cached }
            let result = lexicon.descend(cursor, appendingUTF8: bytes)
            shallowCache[key] = result
            return result
        }

        var heap = BinaryHeap()
        var visited: [VisitKey: Double] = [:]
        var results: [(word: String, cost: Double)] = []
        var emitted = Set<String>()

        let multiEditCap = min(costCap, config.beamMultiEditCostCap)
        func push(_ state: State) {
            let cap = state.edits >= 2 ? multiEditCap : costCap
            guard state.cost <= cap, !state.cursor.isEmpty else { return }
            let key = VisitKey(
                lowerBound: state.cursor.lowerBound,
                byteDepth: state.cursor.byteDepth,
                consumed: state.consumed
            )
            if let best = visited[key], best <= state.cost { return }
            visited[key] = state.cost
            heap.push(state)
        }

        push(
            State(cursor: lexicon.prefixRootCursor(), consumed: 0, cost: 0, edits: 0)
        )

        let deadline = ContinuousClock.now + .seconds(config.beamTimeBudget)
        var expansions = 0
        while let state = heap.pop() {
            // Lazy-deletion check: a cheaper duplicate may have superseded
            // this entry after it was pushed.
            let key = VisitKey(
                lowerBound: state.cursor.lowerBound,
                byteDepth: state.cursor.byteDepth,
                consumed: state.consumed
            )
            if let best = visited[key], best < state.cost { continue }

            if state.consumed == n {
                if let entry = lexicon.exactEntry(in: state.cursor),
                    emitted.insert(entry.word).inserted
                {
                    results.append((entry.word, state.cost))
                    if results.count >= config.beamMaxCandidates { break }
                }
            }

            // Emit-margin early stop: states pop in cost order, so once the
            // frontier sits far above the best emitted candidate, nothing
            // it can still reach will survive the language re-rank (whose
            // swing is bounded by λ·τ·Δz, well under the margin).
            if let cheapest = results.first,
                state.cost > cheapest.cost + config.beamEmitCostMargin
            {
                break
            }

            expansions += 1
            if expansions > config.beamMaxExpansions { break }
            if expansions % 64 == 0, ContinuousClock.now >= deadline { break }

            let editsLeft = state.edits < maxEdits

            // Small ranges expand over their ACTUAL children (one linear
            // scan) instead of probing the alphabet with binary searches —
            // this is what keeps deep states nearly free. Large ranges
            // (near the root) fall back to probing, cushioned by the
            // shallow memo.
            var childMap: [Character: LexiconPrefixCursor]?
            if let children = lexicon.childCursors(
                of: state.cursor, scanLimit: Self.childScanLimit)
            {
                var map: [Character: LexiconPrefixCursor] = [:]
                map.reserveCapacity(children.count)
                for child in children { map[child.character] = child.cursor }
                childMap = map
            }

            func step(_ char: Character, _ bytes: [UInt8]) -> LexiconPrefixCursor? {
                if let childMap { return childMap[char] }
                return descend(state.cursor, char: char, bytes: bytes)
            }

            if state.consumed < n {
                let typedChar = typed[state.consumed]

                // match
                if let next = step(typedChar, typedBytes[state.consumed]) {
                    push(
                        State(
                            cursor: next,
                            consumed: state.consumed + 1,
                            cost: state.cost,
                            edits: state.edits
                        )
                    )
                }

                if editsLeft {
                    // substitution: the first edit may substitute ANY
                    // character up to the cost cap (edits1 parity — far-key
                    // single subs stay suggestible); later edits only the
                    // cheap-neighbor radius (beamNeighborMaxCost). With a
                    // child map the children themselves are iterated and
                    // priced through the provider — the same set, since
                    // membership is decided by the provider cost either
                    // way; without one, the static candidate lists bound
                    // the probes and cost is checked BEFORE each descend.
                    let subCap = state.edits == 0 ? costCap : min(costCap, config.beamNeighborMaxCost + state.cost)
                    if let childMap {
                        for (char, next) in childMap where char != typedChar {
                            let cost =
                                state.cost
                                + substitutionPrice(
                                    position: state.consumed, typedChar: typedChar, intended: char)
                            guard cost <= subCap else { continue }
                            push(
                                State(
                                    cursor: next,
                                    consumed: state.consumed + 1,
                                    cost: cost,
                                    edits: state.edits + 1
                                )
                            )
                        }
                    } else {
                        let candidates =
                            state.edits == 0
                            ? alphabet
                            : cheapNeighbors[typedChar] ?? []
                        for entry in candidates where entry.char != typedChar {
                            let cost =
                                state.cost
                                + substitutionPrice(
                                    position: state.consumed, typedChar: typedChar,
                                    intended: entry.char)
                            guard cost <= subCap else { continue }
                            push(
                                State(
                                    cursor: descend(
                                        state.cursor, char: entry.char, bytes: entry.bytes),
                                    consumed: state.consumed + 1,
                                    cost: cost,
                                    edits: state.edits + 1
                                )
                            )
                        }
                    }

                    // deletion: the typed char was unintended.
                    // (No descend, so the visit key changes only in
                    // `consumed`.)
                    if state.cost + deletionCost <= costCap {
                        push(
                            State(
                                cursor: state.cursor,
                                consumed: state.consumed + 1,
                                cost: state.cost + deletionCost,
                                edits: state.edits + 1
                            )
                        )
                    }

                    // transposition of two adjacent typed chars.
                    if state.consumed + 1 < n,
                        typed[state.consumed] != typed[state.consumed + 1],
                        state.cost + transpositionCost <= costCap,
                        let first = step(typed[state.consumed + 1], typedBytes[state.consumed + 1]),
                        !first.isEmpty
                    {
                        push(
                            State(
                                cursor: descend(
                                    first, char: typedChar, bytes: typedBytes[state.consumed]),
                                consumed: state.consumed + 2,
                                cost: state.cost + transpositionCost,
                                edits: state.edits + 1
                            )
                        )
                    }
                }
            }

            // insertion: the user omitted an intended character — legal at
            // any point INCLUDING after the last typed char (trailing
            // omissions; the DP re-score prices strict-prefix completions
            // down to completionCharCost, mirroring the old edits1).
            // Omitted APOSTROPHES fold under the EN lane profile (dont →
            // don't at ~ε), so their price flows through the lane pricing.
            if editsLeft {
                if let childMap {
                    for (char, next) in childMap {
                        let cost =
                            state.cost + pricing.omissionPrice(of: char, base: insertionCost)
                        guard cost <= costCap else { continue }
                        push(
                            State(
                                cursor: next,
                                consumed: state.consumed,
                                cost: cost,
                                edits: state.edits + 1
                            )
                        )
                    }
                } else {
                    for entry in alphabet {
                        let cost =
                            state.cost
                            + pricing.omissionPrice(of: entry.char, base: insertionCost)
                        guard cost <= costCap else { continue }
                        push(
                            State(
                                cursor: descend(state.cursor, char: entry.char, bytes: entry.bytes),
                                consumed: state.consumed,
                                cost: cost,
                                edits: state.edits + 1
                            )
                        )
                    }
                }
            }
        }
        if Self.debugEnabled {
            let us = start.duration(to: .now)
            print(
                "beam: typed=\(String(typed)) lex=\(lexiconIndex) expansions=\(expansions) "
                    + "visited=\(visited.count) results=\(results.count) t=\(us)")
        }
        return results
    }

    private static let debugEnabled =
        ProcessInfo.processInfo.environment["TYPE_BEAM_DEBUG"] != nil

    /// Minimal binary min-heap on `State.cost` (ties broken arbitrarily —
    /// dedup and the final DP re-score make the order within a cost tie
    /// irrelevant).
    private struct BinaryHeap {
        private var storage: [State] = []

        mutating func push(_ element: State) {
            storage.append(element)
            var child = storage.count - 1
            while child > 0 {
                let parent = (child - 1) >> 1
                guard storage[child].cost < storage[parent].cost else { break }
                storage.swapAt(child, parent)
                child = parent
            }
        }

        mutating func pop() -> State? {
            guard let first = storage.first else { return nil }
            let last = storage.removeLast()
            if !storage.isEmpty {
                storage[0] = last
                var parent = 0
                while true {
                    let left = parent * 2 + 1
                    guard left < storage.count else { break }
                    let right = left + 1
                    var smallest = left
                    if right < storage.count, storage[right].cost < storage[left].cost {
                        smallest = right
                    }
                    guard storage[smallest].cost < storage[parent].cost else { break }
                    storage.swapAt(parent, smallest)
                    parent = smallest
                }
            }
            return first
        }
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/BeamDecoder.kt
package `is`.solberg.lyklabord.engine

import `is`.solberg.lyklabord.engine.config.EngineConfig
import `is`.solberg.lyklabord.engine.lexicon.LexiconPrefixCursor
import `is`.solberg.lyklabord.engine.lexicon.PrefixSearchableLexicon

/** Per-position character-level likelihood seam for beam decoding. */
interface PositionCostProvider {
    /** -log P(typed | intended), in nats; zero when the characters match. */
    fun substitutionCost(position: Int, typed: Char, intended: Char): Double

    /** Confidence that the tap at [position] selected its resolved character. */
    fun confidence(position: Int): Double

    /** Whether [position] has usable per-tap evidence. */
    fun hasTap(position: Int): Boolean = false
}

/** Static key-center provider: no per-tap evidence, confidence is always one. */
class StaticSpatialCostProvider(
    val spatial: SpatialModel,
) : PositionCostProvider {
    override fun substitutionCost(position: Int, typed: Char, intended: Char): Double =
        spatial.substitutionCost(typed, intended)

    override fun confidence(position: Int): Double = 1.0
}

/** A word and its beam-search channel cost, emitted cheapest first. */
data class BeamCandidate(val word: String, val cost: Double)

/**
 * Spatial lexicon beam search. The search walks typed characters and a lexicon
 * prefix cursor in lockstep, visiting only prefixes present in the lexicon.
 */
class BeamDecoder(
    val config: EngineConfig,
    spatial: SpatialModel,
) {
    private data class AlphabetEntry(val char: Char, val bytes: ByteArray)

    /** Full candidate alphabet used for first substitutions and insertions. */
    private val alphabet: List<AlphabetEntry> = candidateAlphabet.map {
        AlphabetEntry(it, it.toString().toByteArray(Charsets.UTF_8))
    }

    /** Cheap static neighbors used for subsequent substitutions. */
    private val cheapNeighbors: Map<Char, List<AlphabetEntry>> =
        candidateAlphabet.associateWith { typed ->
            alphabet.filter { entry ->
                entry.char != typed &&
                    spatial.substitutionCost(typed, entry.char) <= config.beamNeighborMaxCost
            }
        }

    private data class ShallowKey(
        val lexicon: Int,
        val lowerBound: Int,
        val byteDepth: Int,
        val char: Char,
    )

    private val shallowCache = HashMap<ShallowKey, LexiconPrefixCursor>()

    private data class State(
        val cursor: LexiconPrefixCursor,
        val consumed: Int,
        val cost: Double,
        val edits: Int,
    )

    private data class VisitKey(
        val lowerBound: Int,
        val byteDepth: Int,
        val consumed: Int,
    )

    /** Decode [typed] against one lexicon, up to the configured candidate cap. */
    fun decode(
        typed: List<Char>,
        lexicon: PrefixSearchableLexicon,
        lexiconIndex: Int,
        costs: PositionCostProvider,
        pricing: FoldPricing,
        maxEdits: Int,
        multiEditCostCap: Double? = null,
    ): List<BeamCandidate> {
        val n = typed.size
        if (n == 0 || maxEdits < 0) return emptyList()

        val costCap = config.beamCostCap
        val insertionCost = config.spatialCosts.deletion
        val deletionCost = config.spatialCosts.insertion
        val transpositionCost = config.spatialCosts.transposition
        val typedBytes = typed.map { it.toString().toByteArray(Charsets.UTF_8) }
        val foldPenaltyCap = config.tapFoldConfidenceMaxPenalty

        fun substitutionPrice(position: Int, typedChar: Char, intended: Char): Double {
            val base = costs.substitutionCost(position, typedChar, intended)
            val lane = pricing.substitutionPrice(typedChar, intended, confusionBase = base)
                ?: return base
            val evidence = costs.foldEvidencePenalty(position, foldPenaltyCap)
            return minOf(lane + evidence, base)
        }

        fun descend(
            cursor: LexiconPrefixCursor,
            char: Char,
            bytes: ByteArray,
        ): LexiconPrefixCursor {
            if (cursor.byteDepth > 2 || cursor.isEmpty) {
                return lexicon.descend(cursor, bytes)
            }
            val key = ShallowKey(lexiconIndex, cursor.lowerBound, cursor.byteDepth, char)
            shallowCache[key]?.let { return it }
            val result = lexicon.descend(cursor, bytes)
            shallowCache[key] = result
            return result
        }

        val startNanos = System.nanoTime()
        val heap = StateHeap()
        val visited = HashMap<VisitKey, Double>()
        val results = ArrayList<BeamCandidate>()
        val emitted = HashSet<String>()
        val multiEditCap = minOf(costCap, multiEditCostCap ?: config.beamMultiEditCostCap)

        fun push(state: State) {
            val cap = if (state.edits >= 2) multiEditCap else costCap
            if (state.cost > cap || state.cursor.isEmpty) return
            val key = VisitKey(state.cursor.lowerBound, state.cursor.byteDepth, state.consumed)
            val best = visited[key]
            if (best != null && best <= state.cost) return
            visited[key] = state.cost
            heap.push(state)
        }

        push(State(lexicon.prefixRootCursor(), consumed = 0, cost = 0.0, edits = 0))

        val deadline = startNanos + (config.beamTimeBudget * 1_000_000_000.0).toLong()
        var expansions = 0
        while (true) {
            val state = heap.pop() ?: break
            val key = VisitKey(state.cursor.lowerBound, state.cursor.byteDepth, state.consumed)
            val best = visited[key]
            if (best != null && best < state.cost) continue

            if (state.consumed == n) {
                val entry = lexicon.exactEntry(state.cursor)
                if (entry != null && emitted.add(entry.word)) {
                    results += BeamCandidate(entry.word, state.cost)
                    if (results.size >= config.beamMaxCandidates) break
                }
            }

            val cheapest = results.firstOrNull()
            if (cheapest != null && state.cost > cheapest.cost + config.beamEmitCostMargin) break

            expansions++
            if (expansions > config.beamMaxExpansions) break
            if (expansions % 64 == 0 && System.nanoTime() >= deadline) break

            val editsLeft = state.edits < maxEdits
            val childList = lexicon.childCursors(state.cursor, childScanLimit)
            val childMap = childList?.associate { it.character to it.cursor }
            fun step(char: Char, bytes: ByteArray): LexiconPrefixCursor? =
                if (childMap != null) childMap[char]
                else descend(state.cursor, char, bytes).takeUnless { it.isEmpty }

            if (state.consumed < n) {
                val typedChar = typed[state.consumed]
                val match = step(typedChar, typedBytes[state.consumed])
                if (match != null) {
                    push(State(match, state.consumed + 1, state.cost, state.edits))
                }

                if (editsLeft) {
                    val subCap = if (state.edits == 0) {
                        costCap
                    } else {
                        minOf(costCap, config.beamNeighborMaxCost + state.cost)
                    }
                    if (childList != null) {
                        for (child in childList) {
                            if (child.character == typedChar) continue
                            val cost = state.cost + substitutionPrice(
                                state.consumed, typedChar, child.character,
                            )
                            if (cost > subCap) continue
                            push(State(child.cursor, state.consumed + 1, cost, state.edits + 1))
                        }
                    } else {
                        val candidates = if (state.edits == 0) {
                            alphabet
                        } else {
                            cheapNeighbors[typedChar].orEmpty()
                        }
                        for (entry in candidates) {
                            if (entry.char == typedChar) continue
                            val cost = state.cost + substitutionPrice(
                                state.consumed, typedChar, entry.char,
                            )
                            if (cost > subCap) continue
                            push(
                                State(
                                    descend(state.cursor, entry.char, entry.bytes),
                                    state.consumed + 1,
                                    cost,
                                    state.edits + 1,
                                ),
                            )
                        }
                    }

                    if (state.cost + deletionCost <= costCap) {
                        push(
                            State(
                                state.cursor,
                                state.consumed + 1,
                                state.cost + deletionCost,
                                state.edits + 1,
                            ),
                        )
                    }

                    if (
                        state.consumed + 1 < n &&
                        typed[state.consumed] != typed[state.consumed + 1] &&
                        state.cost + transpositionCost <= costCap
                    ) {
                        val first = step(typed[state.consumed + 1], typedBytes[state.consumed + 1])
                        if (first != null) {
                            val second = descend(
                                first,
                                typedChar,
                                typedBytes[state.consumed],
                            )
                            push(
                                State(
                                    second,
                                    state.consumed + 2,
                                    state.cost + transpositionCost,
                                    state.edits + 1,
                                ),
                            )
                        }
                    }
                }
            }

            if (editsLeft) {
                if (childList != null) {
                    for (child in childList) {
                        val cost = state.cost + pricing.omissionPrice(
                            child.character,
                            base = insertionCost,
                        )
                        if (cost > costCap) continue
                        push(State(child.cursor, state.consumed, cost, state.edits + 1))
                    }
                } else {
                    for (entry in alphabet) {
                        val cost = state.cost + pricing.omissionPrice(
                            entry.char,
                            base = insertionCost,
                        )
                        if (cost > costCap) continue
                        push(
                            State(
                                descend(state.cursor, entry.char, entry.bytes),
                                state.consumed,
                                cost,
                                state.edits + 1,
                            ),
                        )
                    }
                }
            }
        }

        if (debugEnabled) {
            val elapsed = (System.nanoTime() - startNanos) / 1_000_000.0
            println(
                "beam: typed=${typed.joinToString("")} lex=$lexiconIndex " +
                    "expansions=$expansions visited=${visited.size} " +
                    "results=${results.size} t=${elapsed}ms",
            )
        }
        return results
    }

    private class StateHeap {
        private val storage = ArrayList<State>()

        fun push(element: State) {
            storage += element
            var child = storage.lastIndex
            while (child > 0) {
                val parent = (child - 1) shr 1
                if (storage[child].cost >= storage[parent].cost) break
                storage.swap(child, parent)
                child = parent
            }
        }

        fun pop(): State? {
            if (storage.isEmpty()) return null
            val first = storage[0]
            val last = storage.removeAt(storage.lastIndex)
            if (storage.isNotEmpty()) {
                storage[0] = last
                var parent = 0
                while (true) {
                    val left = parent * 2 + 1
                    if (left >= storage.size) break
                    val right = left + 1
                    var smallest = left
                    if (right < storage.size && storage[right].cost < storage[left].cost) {
                        smallest = right
                    }
                    if (storage[smallest].cost >= storage[parent].cost) break
                    storage.swap(parent, smallest)
                    parent = smallest
                }
            }
            return first
        }

        private fun ArrayList<State>.swap(a: Int, b: Int) {
            val value = this[a]
            this[a] = this[b]
            this[b] = value
        }
    }

    private companion object {
        const val childScanLimit = 48
        val candidateAlphabet: List<Char> =
            "aábcdðeéfghiíjklmnoópqrstuúvwxyýzþæö'’".toList()
        val debugEnabled: Boolean = System.getenv("TYPE_BEAM_DEBUG") != null
    }
}
